tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,252 @@
+                    {
+                        foundFields[j] = field;
+                        found++;
+                        break;
+                    }
+                }
+            }
+
+            return foundFields;
+        }
+
+        /// <summary>
+        /// MethodMetadataInfo saves locals from MoveNext's localVarSig,
+        /// This isn't enough in the async scenario because we need to extract more locals that may have been hoisted in the builder object,
+        /// and we need to subtract some locals that exist in the localVarSig but they do not belongs to the kickoff method
+        /// For now, we're capturing all locals that are hoisted (except known generated locals),
+        /// and we're capturing all the locals that appear in localVarSig with ` LogLocal `
+        /// </summary>
+        /// <typeparam name="TTarget">Instance type</typeparam>
+        /// <param name="instance">Instance object</param>
+        /// <param name="asyncKickoffInfo">The async kickoff info</param>
+        /// <returns>List of locals</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static FieldInfoNameSanitized[] GetHoistedLocalsFromStateMachine<TTarget>(TTarget instance, AsyncKickoffMethodInfo asyncKickoffInfo)
+        {
+            var foundFields = new List<FieldInfoNameSanitized>();
+            var allFields = instance.GetType().GetFields(AllFieldsBindingFlags);
+            var kickoffParameters = asyncKickoffInfo.KickoffMethod.GetParameters();
+            for (var i = 0; i < allFields.Length; i++)
+            {
+                var field = allFields[i];
+                if (IsStateMachineField(field))
+                {
+                    continue;
+                }
+
+                bool foundInArguments = false;
+                for (int j = 0; j < kickoffParameters.Length; j++)
+                {
+                    if (kickoffParameters[j].Name == field.Name)
+                    {
+                        foundInArguments = true;
+                        break;
+                    }
+                }
+
+                if (foundInArguments == false)
+                {
+                    var indexOfStateMachineSuffix = field.Name.IndexOf(StateMachineNameSuffix, StringComparison.Ordinal);
+                    if (indexOfStateMachineSuffix > 0)
+                    {
+                        foundFields.Add(new FieldInfoNameSanitized(field, SanitizeAsyncHoistedLocalName(field.Name, indexOfStateMachineSuffix)));
+                    }
+                    else
+                    {
+                        foundFields.Add(new FieldInfoNameSanitized(field, field.Name));
+                    }
+                }
+            }
+
+            return foundFields.ToArray();
+        }
+
+        private static string SanitizeAsyncHoistedLocalName(string fieldName, int indexOfStateMachineSuffix)
+        {
+#if NETCOREAPP3_0_OR_GREATER
```

Comment:
Seems like a bug. Don't we need to make this an `if statement` rather than a preprocessor directive, since the `netstandard2.0` version of `Datadog.Trace.DLL` could be used in both `netcoreapp3.0` and `net5`, for example?

If I'm right and it is a bug, maybe worthwhile to understand why our tests did not pick up on it? Is this an untested code path or does it have something to do with how our tests work?

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26